lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+    OpBuilder::InsertionGuard guard(builder);
+    if (!operand.isa<BlockArgument>() && operand.getDefiningOp()) {
+      builder.setInsertionPointAfter(operand.getDefiningOp());
+    }
+    auto castOp = builder.create<arith::IndexCastOp>(
+        loc, builder.getIndexType(), operand);
+    state.scalar = castOp.getResult();
+    return success();
+  } else if (operand.getType().isa<IndexType>()) {
+    state.scalar = operand;
+    return success();
+  }
+
+  if (operand.getType().isa<triton::PointerType>()) {
+    // A scalar pointer can either be produced by AddPtrOp or a block
+    // argument
+    if (auto op = operand.getDefiningOp()) {
+      if (auto addPtrOp = dyn_cast<triton::AddPtrOp>(op)) {
+        return visitOperandAddptr(cast<triton::AddPtrOp>(op), state, loc,
+                                  builder);
+      } else if (auto makeTensorOp = dyn_cast<triton::MakeTensorPtrOp>(op)) {
+        llvm_unreachable("NYI");
+      } else {
+        llvm_unreachable("Unexpected operand defining operation");
+      }
+    } else {
+      state.source = operand;
+      return success();
+    }
+  }
+
+  if (auto op = operand.getDefiningOp<arith::AddIOp>()) {
+    return visitOperandAdd(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<arith::MulIOp>()) {
+    return visitOperandMul(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::MakeRangeOp>()) {
+    return visitOperandMakeRange(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::BroadcastOp>()) {
+    return visitOperandBroadcast(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::SplatOp>()) {
+    return visitOperandSplat(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::ExpandDimsOp>()) {
+    return visitOperandExpandDims(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<triton::AddPtrOp>()) {
+    return visitOperandAddptr(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<arith::ConstantOp>()) {
+    return visitOperandConstSplat(op, state, loc, builder);
+  } else if (auto op = operand.getDefiningOp<arith::RemSIOp>()) {
+    return visitOperandRem(op, state, loc, builder);
+  } else {
+    llvm::dbgs() << "PtrAnalysis: encountered addptr operand produced by an "
+                    "unsupported operation\n";
+    operand.dump();
+    return failure();
+  }
+}
+
+LogicalResult PtrAnalysis::rewriteAddptrOp(triton::AddPtrOp op) {
+  OpBuilder builder(op);
+
+  PtrSState state;
+  if (visitOperandAddptr(op, state, op.getLoc(), builder).failed()) {
+    return failure();
+  }
+
+  knownPtrs[op.getResult()] = state;
+
+  if (op.getPtr().getType().isa<RankedTensorType>()) {
+    auto maketptrOp = state.createTTSMakeTensorPtrOp(builder, op.getLoc());
+    map.map(op.getResult(), maketptrOp.getResult());
+  } else {
+    map.map(op.getResult(), op.getResult());
+  }
```

Comment:
is this map redundant since it's just mapping to itself?

Reply:
oh i see we're using this map in `rewriteForOp` to also check if a value is pointer related

Reply:
Yes exactly. Will add a comment.